### PR TITLE
Scripts to run the examples should not modify the cache

### LINKED ISSUE #28: Do not change conan client configuration permanently when running examples
Some examples like the one of the lockfiles, changes the Conan client configuration with ``conan config set ...``. Running the example script locally leaves your configuration with that setting. Instead of this, examples should set the configuration via env vars temporarily just for the run of that example.

### PR DESCRIPTION
**Pull request by jgsogo**
The examples are run using scripts like `build.py`, `build.bat` or `build.sh`, these scripts are inside the folder where the example is contained and users may run them to check that example works in their machine. Taken into account this, examples can't modify the user cache (they may install packages but configuration files, remotes,... everything should be the same before and after).

Not covered in this PR: metadata of references, we don't want an example to change the `remote` associated with a reference or package.

closes #28

### CODE REVIEW

File: .ci/run.py [hunk 1]
```diff
@@ -96,6 +99,48 @@ def print_build(script):
     writeln_console("================================================================")
 
 
+@contextmanager
+def ensure_cache_preserved():
```
Comment (uilianries) [inline .ci/run.py]: clever solution! I would use git init + git add + git diff to check any different, but your solution works pretty well :)

File: .ci/run.py [hunk 2]
```diff
@@ -105,10 +150,15 @@ def run_scripts(scripts):
...
             print_build(script)
-            if abspath.endswith(".py"):
-                result = subprocess.call([sys.executable, abspath], env=env)
-            else:
-                result = subprocess.call(abspath, env=env)
+            build_script = [sys.executable, abspath] if abspath.endswith(".py") else abspath
+            
+            # Need to initialize the cache with default files if they are not already there
+            subprocess.call(['conan', 'install', 'zlib/1.2.11@conan/stable'], env=env)
+            subprocess.call(['conan', 'config', 'set', 'log.print_run_commands=True'], env=env)
```
Comment (uilianries) [inline .ci/run.py]: why not enabling it in Travis CI? you could set `CONAN_PRINT_RUN_COMMANDS=1`

File: features/lockfiles/ci/build.py [hunk 3]
```diff
@@ -15,53 +18,71 @@ def rm(path):
...
-    _, pkg_ref = build_order[0][0]
-    pkg_ref = pkg_ref.split("#", 1)[0]
-    print("\n********** Rebuild affected package: %s ***********" % pkg_ref)
-    run("conan install %s --build=%s --lockfile=release" % (pkg_ref, pkg_ref))
-    os.chdir("..")
-    run("conan graph update-lock release build_server_folder/release")
+
+@contextmanager
+def restore_conan_home():
+    conan_home = os.environ.get('CONAN_USER_HOME', None) or os.path.expanduser('~')
```
Comment (uilianries) [inline features/lockfiles/ci/build.py]:
```suggestion
    conan_home = os.environ.get('CONAN_USER_HOME', os.path.expanduser('~'))
```

**Comment by uilianries** [comment]
I know it's a draft, but the idea sounds good :+1:

File: .ci/run.py (hunk 1 above)
Comment (jgsogo) [reply .ci/run.py]: I think I like yours more. Easier to implement, cleaner... it is just much better. Thanks!

Comment (uilianries) [reply .ci/run.py]: oh, much simpler now!

**Comment by jgsogo** [comment]
It was a draft, but from my POV is ready to merge, right?

**Comment by danimtb** [comment]
I think it makes sense and it is safe to merge

**Comment by jgsogo** [comment]
My only concern is the `subprocess.call(['conan', 'install', 'zlib/1.2.11@conan/stable'], env=env)` line, I would love to remove it, but I need to initialize the `settings.yml`, `remotes.json`, `profile-default`,... all those files that are automagically created if they are not already there. If anyone thinks about an alternative, please, say, it would be needed.

**Comment by uilianries** [comment]
is possible using some wrong reference? you could try:

```python
try:
    subprocess.call(['conan', 'install', 'foobar/0.1.0@conan/stable'], env=env)
except:
    pass
```

**Comment by danimtb** [comment]
What about just ``conan profile new default --detect``? I think it initializes everything additionally to creating the default profile

**Comment by jgsogo** [comment]
Suggestion by @uilianries, works. If I just run `conan profile. ...` it doesn't initialize the remotes :/

**Comment by uilianries** [comment]
It looks nice, ready to go IMO

**Pull request merged**